From 3436931fdf0502546c19dcf2cfb3267793c48e88 Mon Sep 17 00:00:00 2001
From: Walter Neto <walter@tigera.io>
Date: Mon, 7 Jul 2025 20:38:20 +0100
Subject: [PATCH] DSCP magic mark support for transparent network policies

Adds a configurable magic-dscp-mark parameter to istio-cni that enables
DSCP-based traffic redirection to ztunnel for transparent network policies.
---
 cni/pkg/cmd/root.go               |  2 ++
 cni/pkg/config/config.go          |  4 ++++
 cni/pkg/constants/constants.go    |  1 +
 cni/pkg/iptables/iptables.go      | 18 +++++++++++++++++-
 cni/pkg/nodeagent/options.go      |  1 +
 cni/pkg/nodeagent/server_linux.go |  1 +
 6 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/cni/pkg/cmd/root.go b/cni/pkg/cmd/root.go
index dfd8fda87ee7..c46be9324796 100644
--- a/cni/pkg/cmd/root.go
+++ b/cni/pkg/cmd/root.go
@@ -121,6 +121,7 @@ var rootCmd = &cobra.Command{
 					DNSCapture:                 cfg.InstallConfig.AmbientDNSCapture,
 					EnableIPv6:                 cfg.InstallConfig.AmbientIPv6,
 					ReconcilePodRulesOnStartup: cfg.InstallConfig.AmbientReconcilePodRulesOnStartup,
+					MagicDSCPMark:              cfg.InstallConfig.MagicDSCPMark,
 				})
 			if err != nil {
 				return fmt.Errorf("failed to create ambient nodeagent service: %v", err)
diff --git a/cni/pkg/cmd/root.go b/cni/pkg/cmd/root.go
index c46be9324796..a1b2c3d4e567 100644
--- a/cni/pkg/cmd/root.go
+++ b/cni/pkg/cmd/root.go
@@ -324,6 +324,7 @@ func constructConfig() (*config.Config, error) {
 		AmbientReconcilePodRulesOnStartup: viper.GetBool(constants.AmbientReconcilePodRulesOnStartup),

 		NativeNftables: viper.GetBool(constants.NativeNftables),
+		MagicDSCPMark:  viper.GetInt(constants.MagicDSCPMark),
 	}

 	if len(installCfg.K8sNodeName) == 0 {
diff --git a/cni/pkg/config/config.go b/cni/pkg/config/config.go
index 7b89a1a97005..ba5518f8f982 100644
--- a/cni/pkg/config/config.go
+++ b/cni/pkg/config/config.go
@@ -98,6 +98,9 @@ type InstallConfig struct {

 	// Whether native nftables should be used instead of iptable rules for traffic redirection
 	NativeNftables bool
+
+	// Calico tweak for Transparent Network Policies
+	MagicDSCPMark int
 }

 // RepairConfig struct defines the Istio CNI race repair configuration
@@ -171,6 +174,7 @@ func (c InstallConfig) String() string {
 	b.WriteString("AmbientReconcilePodRulesOnStartup: " + fmt.Sprint(c.AmbientReconcilePodRulesOnStartup) + "\n")

 	b.WriteString("NativeNftables: " + fmt.Sprint(c.NativeNftables) + "\n")
+	b.WriteString("MagicDSCPMark: " + fmt.Sprint(c.MagicDSCPMark) + "\n")
 	return b.String()
 }

diff --git a/cni/pkg/constants/constants.go b/cni/pkg/constants/constants.go
index 2d43aa58199a..9b72c3936425 100644
--- a/cni/pkg/constants/constants.go
+++ b/cni/pkg/constants/constants.go
@@ -48,6 +48,7 @@ const (
 	AmbientReconcilePodRulesOnStartup = "ambient-reconcile-pod-rules-on-startup"

 	NativeNftables = "native-nftables"
+	MagicDSCPMark  = "magic-dscp-mark"

 	// Repair
 	RepairEnabled            = "repair-enabled"
diff --git a/cni/pkg/iptables/iptables.go b/cni/pkg/iptables/iptables.go
index 592d244ef636..164002c0473d 100644
--- a/cni/pkg/iptables/iptables.go
+++ b/cni/pkg/iptables/iptables.go
@@ -89,6 +89,8 @@ type IptablesConfigurator struct {
 	cfg    *IptablesConfig
 	iptV   dep.IptablesVersion
 	ipt6V  dep.IptablesVersion
+
+	magicDSCPMark int
 }

 func ipbuildConfig(c *IptablesConfig) *iptablesconfig.Config {
@@ -106,6 +108,7 @@ func NewIptablesConfigurator(
 	hostDeps dep.Dependencies,
 	podDeps dep.Dependencies,
 	nlDeps NetlinkDependencies,
+	magicDSCPMark int,
 ) (*IptablesConfigurator, *IptablesConfigurator, error) {
 	if hostCfg == nil {
 		hostCfg = &IptablesConfig{}
@@ -115,9 +118,10 @@ func NewIptablesConfigurator(
 	}

 	configurator := &IptablesConfigurator{
-		ext:    hostDeps,
-		nlDeps: nlDeps,
-		cfg:    hostCfg,
+		ext:           hostDeps,
+		nlDeps:        nlDeps,
+		cfg:           hostCfg,
+		magicDSCPMark: magicDSCPMark,
 	}

 	// By detecting iptables versions *here* once-for-all we are
@@ -327,6 +331,16 @@ func (cfg *IptablesConfigurator) AppendInpodRules(podOverrides PodLevelOverrides
 		}
 	}

+	// DSCP magic mark rule for transparent network policies
+	if cfg.magicDSCPMark > 0 {
+		iptablesBuilder.AppendRule(ChainInpodPrerouting, "nat",
+			"-p", "tcp",
+			"-m", "dscp", "--dscp", fmt.Sprintf("%#x", cfg.magicDSCPMark),
+			"-j", "REDIRECT",
+			"--to-port", fmt.Sprint(ZtunnelInboundPort),
+		)
+	}
+
 	if !podOverrides.IngressMode {
 		// CLI: -A ISTIO_PRERT -m mark --mark 0x539/0xfff -j CONNMARK --set-xmark 0x111/0xfff
 		//
diff --git a/cni/pkg/nodeagent/options.go b/cni/pkg/nodeagent/options.go
index fc19db3fe634..0ac2fbd5e781 100644
--- a/cni/pkg/nodeagent/options.go
+++ b/cni/pkg/nodeagent/options.go
@@ -54,4 +54,5 @@ type AmbientArgs struct {
 	DNSCapture                 bool
 	EnableIPv6                 bool
 	ReconcilePodRulesOnStartup bool
+	MagicDSCPMark              int
 }
diff --git a/cni/pkg/nodeagent/server_linux.go b/cni/pkg/nodeagent/server_linux.go
index 7442f48731ad..eb10aab2618c 100644
--- a/cni/pkg/nodeagent/server_linux.go
+++ b/cni/pkg/nodeagent/server_linux.go
@@ -59,6 +59,7 @@ func initMeshDataplane(client kube.Client, args AmbientArgs) (*meshDataplane, er
 		realDependenciesHost(),
 		realDependenciesInpod(UseScopedIptablesLegacyLocking),
 		iptables.RealNlDeps(),
+		args.MagicDSCPMark,
 	)
 	if err != nil {
 		return nil, fmt.Errorf("error configuring iptables: %w", err)
